backend-rust/src/graphql_api.rs

```diff
@@ -1812,8 +1818,93 @@ impl Contract {
         todo_api!()
     }
 
-    async fn tokens(&self, skip: u32, take: u32) -> ApiResult<TokensCollectionSegment> {
-        todo_api!()
+    // This function fetches CIS2 tokens associated to a given contract, ordered by
+    // their creation index in descending order. It retrieves the most recent
+    // tokens first, with support for pagination through `skip` and `take`
+    // parameters.
+    // - `skip` determines how many of the most recent tokens to skip.
+    // - `take` controls the number of tokens to return, respecting the collection
+    //   limit.
+    async fn tokens(
+        &self,
+        ctx: &Context<'_>,
+        skip: Option<u64>,
+        take: Option<u64>,
+    ) -> ApiResult<TokensCollectionSegment> {
+        let config = get_config(ctx)?;
+        let pool = get_pool(ctx)?;
+
+        let max_token_index = sqlx::query!(
+            "SELECT MAX(token_index_per_contract) AS max_token_index
+            FROM tokens
+            WHERE tokens.contract_index = $1 AND tokens.contract_sub_index = $2",
+            self.contract_address_index.0 as i64,
+            self.contract_address_sub_index.0 as i64,
+        )
+        .fetch_one(pool)
+        .await?
+        .max_token_index
+        .unwrap_or(0) as u64;
```

Comment:
Minor thing: For queries for just one value, we can use `query_scalar`
```suggestion
        let max_token_index = sqlx::query_scalar!(
            "SELECT MAX(token_index_per_contract)
            FROM tokens
            WHERE tokens.contract_index = $1 AND tokens.contract_sub_index = $2",
            self.contract_address_index.0 as i64,
            self.contract_address_sub_index.0 as i64,
        )
        .fetch_one(pool)
        .await?
        .unwrap_or(0) as u64;
```